Significantly reduce the number of redundant parentheses in the generated code

izeigerman opened the pull request:
Admittedly the original idea of wrapping each operand of an infix expression into parentheses was a half-assed solution. 
I've noticed recurrent complains from users who mentioned bloated file size, crashed editors, etc. So I figured I try and introduce the proper operator precedence to the code generation logic and only add parentheses where they are necessary.

izeigerman commented:
Well, actually, the infix power operator seems to be common enough to deserve a mixin of its own or some other means of generalization. Will refactor this.

izeigerman commented:
@StrikerRUS I think this should be ready for review now. Thanks!

izeigerman commented:
I'm going to regenerate code examples in a follow-up PR.

StrikerRUS commented:
I'm so excited you've come back to this project!

Also I would like to sorry for not allocating significant amount of time and energy to this project recently.

StrikerRUS commented on m2cgen/interpreters/interpreter.py:
```diff
@@ -190,13 +201,10 @@ def interpret_tanh_expr(self, expr, **kwargs):
-        exp_result = self._do_interpret(expr.exp_expr, **kwargs)
-        return self._cg.function_invocation(self.power_function_name, base_result, exp_result)
+    def _wrap_infix_expr(self, expr, left_precedence, right_precedence):
+        wrap = left_precedence is not None and left_precedence > expr.precedence
+        wrap |= right_precedence is not None and right_precedence >= expr.precedence
```
Maybe we can avoid `is not None` by changing default precedence (see comment above)?

StrikerRUS commented on m2cgen/interpreters/mixins.py:
```diff
@@ -195,3 +195,46 @@ def _get_subroutine_name(self):
+
+    pow_operator = "**"
+
+    infix_expressions = [*BaseToCodeInterpreter.infix_expressions, ast.PowExpr]
```
See comment above about eliminating `list -> tuple` cast for some speedup.
```suggestion
    infix_expressions = (*BaseToCodeInterpreter.infix_expressions, ast.PowExpr)
```

izeigerman replied:
I actually didn't know this syntax was supported! Otherwise I wouldn't have introduced the list->tuple conversion. TIL!

StrikerRUS commented on m2cgen/interpreters/mixins.py:
```diff
@@ -195,3 +195,46 @@ def _get_subroutine_name(self):
+            expr.base_expr, left_precedence=expr.precedence, **kwargs)
+        exp_result = self._do_interpret(
+            expr.exp_expr, right_precedence=expr.precedence, **kwargs)
+        return self._cg.infix_expression(
+            left=base_result, right=exp_result, op=self.pow_operator,
+            wrap=self._wrap_infix_expr(expr, left_precedence, right_precedence))
```
Same as above about code readability.
```suggestion
        return self._cg.infix_expression(
            left=base_result,
            right=exp_result,
            op=self.pow_operator,
            wrap=self._wrap_infix_expr(expr, left_precedence, right_precedence))
```

izeigerman replied:
Yeah, I feel like we need to keep up with modern best practices and introduce auto-formatting to the project (eg. `black`). Will address this manually for now and will introduce `black` support in future PRs.

StrikerRUS commented on m2cgen/interpreters/mixins.py:
```diff
@@ -195,3 +195,46 @@ def _get_subroutine_name(self):
         raise NotImplementedError
+
+
+class InfixPowExprMixin(BaseToCodeInterpreter):
```
IMHO, `PowExprInfixMixin` and `PowExprFunctionMixin` names are slightly better because the first word highlights a context about what these classes are (they are about for to implement power expression).

Also, it would be easier for contributors to search for appropriate mixin for a new programming language.

Just as an example: https://docs.opencv.org/3.4/d8/d01/group__imgproc__color__conversions.html

Note enumerations are in a form of `COLOR_xxx` but not in a form of `xxx_COLOR`.

StrikerRUS commented on m2cgen/interpreters/mixins.py:
```diff
@@ -195,3 +195,46 @@ def _get_subroutine_name(self):
+            wrap=self._wrap_infix_expr(expr, left_precedence, right_precedence))
+
+
+class FunctionPowExprMixin(BaseToCodeInterpreter):
```
Maybe make one of these two mixins as a default behavior? This will allow contributors to not get familiar with mixins mechanism in a half (roughly, of course) of all cases.

izeigerman replied:
So I actually did this intentionally :) I want the developer to make a conscious choice and pick one or the other. 

> Maybe make one of these two mixins as a default behavior? 

Well, I don't think there is such thing as a "default" behavior in this case, since even with the pow expression as a function a potential contributor needs to provide a value for the `pow_function_name` which doesn't have a default. So I'd rather prefer this choice to be explicit.

StrikerRUS replied:
OK, fine with me.

StrikerRUS reviewed (changes requested):
Just awesome simplification! Really love it!
Please consider checking some my minor comments below.

izeigerman commented:
> Also I would like to sorry for not allocating significant amount of time and energy to this project recently.

@StrikerRUS you have absolutely nothing to be sorry about! I really appreciate all the time and effort you've put into this project. Thank you and thank you for the great review as always 👍  Working through your comments now

izeigerman commented on m2cgen/interpreters/code_generator.py:
```diff
@@ -99,10 +99,11 @@ def prepend_code_lines(self, lines):
 
-    def infix_expression(self, left, right, op):
-        return self.tpl_infix_expression(left=left, right=right, op=op)
+    # Following methods simply compute expressions using templates without changing result.
```
Ah, right, I totally forgot the context behind that comment. Thanks for reminding me.

izeigerman commented on m2cgen/ast.py:
```diff
@@ -7,6 +7,10 @@
     output_size = 1
+
+    # The precedence of this expression. Higher value means higher precedence.
+    precedence = None
```
Ah, I'm somewhat unsure about this one. On one hand I agree that this simplifies the code quite a bit. On the other hand, I'm hesitant to introduce `precedence` for expressions to which this concept doesn't even apply. `None` in this case means "not applicable", while 0 represents an actual precedence. I feel like I'd like to preserve this distinction and keep it explicit to avoid unexpected behavior / unmet expectations in future. What do you think?

StrikerRUS replied:
> I'm hesitant to introduce `precedence` for expressions to which this concept doesn't even apply.

Makes sense! After your clarification I'd like to preserve `None` as default value too 😃

izeigerman commented:
@StrikerRUS I think I've addressed all your comments at this point except for `precedence = 0` being default for all expressions and the `PowExprFunctionMixin` as a default behavior (see my replies in respective threads). Thanks!

StrikerRUS commented:
> I'm going to regenerate code examples in a follow-up PR.

Let me automate the process of generating code examples. I believe we can make GitHub Actions generate code examples for each commit and pack them into a downloadable artifact. Then, if needed, we can download them and update `master` branch in a separate commit/PR. I can work on this if you don't mind.

StrikerRUS commented on tests/interpreters/test_java.py:
```diff
@@ -288,10 +288,10 @@ def test_depth_threshold_with_bin_expr():
     }
     public static double subroutine0(double[] input) {
-        return (1.0) + ((1.0) + (1.0));
+        return 1.0 + (1.0 + 1.0);
```
Sorry, I can't get it. Why do we have braces here (and in some other places like `var0 = (1.0 + 1.0);`)? Is it due to GE here?
```
wrap |= right_precedence is not None and right_precedence >= expr.precedence
```
https://github.com/BayesWitnesses/m2cgen/pull/511/files#diff-5a4f9c32da5612f71e48dad842ece55b645da127a61b6df8dee1c03ad3db76f1R208

izeigerman replied:
That's an excellent observation!
> Is it due to GE here?

You're absolutely right. Let me share some additional context.
The expression of kind:
```
1.0 + (1.0 + 1.0)
```
is produced as a result of the AST that looks like following:
```
ast.BinNumExpr(
    left=ast.NumVal(1.0),
    right=ast.BinNumExpr(
        left=ast.NumVal(1.0),
        right=ast.NumVal(1.0),
        op=ast.BinNumOpType.ADD
    ),
    op=ast.BinNumOpType.ADD
)
```
If the expression looked like this:
```
ast.BinNumExpr(
    left=ast.BinNumExpr(
        left=ast.NumVal(1.0),
        right=ast.NumVal(1.0),
        op=ast.BinNumOpType.ADD
    ),
    right=ast.NumVal(1.0),
    op=ast.BinNumOpType.ADD
)
```
You would've seen what you expected:
```
1.0 + 1.0 + 1.0
```
It may seem like this kind of difference is weird and it is for associative operations like addition or multiplication, but consider a slightly modified version of the first tree:
```
```
ast.BinNumExpr(
    left=ast.NumVal(1.0),
    right=ast.BinNumExpr(
        left=ast.NumVal(1.0),
        right=ast.NumVal(1.0),
        op=ast.BinNumOpType.SUB
    ),
    op=ast.BinNumOpType.SUB
)
```
This AST produces the following:
```
1.0 - (1.0 - 1.0)
```
Which makes a lot more sense now - omitting parenthesis would've been completely wrong in this case.

We can still get rid of the remaining parenthesis if we optimize our trees to look like 2nd tree rather than the 1st one for associative operation. Or we can introduce some sort of `is_associative` flag to our expressions which I found to be rather redundant. 

Let me know what you think.

izeigerman replied:
Another alternative is to instead of having just `precedence` for each expression we can instead track `left_precedence` and `right_precedence` for each expression individually and get rid of `>=` comparison in the code snippet that you pointed out. 
For associative operators the `left_precedence` and `right_precedence` would have the same value, while for non-associative ones the `right_precedence` will be lower than the `left_precedence`.

izeigerman replied:
Now when I thought about it more I really like the latter approach because it also generalizes well to right associative operators (like eg. `**` in Haskell). Let me know if you have any concerns and if not I'll update the implementation.

izeigerman replied:
Ah, turns out my idea doesn't really cover scenarios like
```
1.0 - (1.0 + 1.0)
```
Please ignore the above comment.

StrikerRUS replied:
Thank you so much for the detailed explanation with excellent examples! The situation is much clearer now for me. I totally support the idea of introducing some kind of associative rule tracking. But I'd like to propose making further updates in a follow-up PR. This PR is already quite big, self-contained and has already eliminate the great amount of excess braces.

izeigerman replied:
Thanks! Merging this then.

StrikerRUS reviewed (approved):
Really great work! Thank you so much! I don't have any other suggestions except one non-blocking comment below. LGTM!
